#!/usr/bin/env python
from dynamic_reconfigure.parameter_generator_catkin import *
PACKAGE = "wallfollowing5"

gen = ParameterGenerator()

gen.add(
    "usable_laser_range",               # name
    double_t,                           # type
    0,                                  # config level
    "Specifies the laser range that is used in degree.",        # description
    240,                                # default
    0.0001,                             # min
    270                                 # max
)

gen.add(
    "target_method",
    int_t,
    0,
    "Defines the used method to dermine the target position - 0: TRACK_CENTER, 1: CIRCLE_TANGENTS, 2: CENTER_PATH",
    2,
    0,
    2
)

gen.add(
    "use_voxel",
    bool_t,
    0,
    "Enable the use of voxel or the direct lidar points.",
    False
)

gen.add(
    "safety_wall_distance",
    double_t,
    0,
    "Specifies the min distance to the walls which shouldn't be undercut.",
    0.3,
    0.0001,
    2
)

gen.add(
    "max_predicted_distance",
    double_t,
    0,
    "Specifies the max predicted distance to determine the target position.",
    1.0,
    0.0001,
    10
)

gen.add(
    "emergency_slowdown",
    bool_t,
    0,
    "Enable the emergency slowdown which is activated in steep curves.",
    True
)

gen.add(
    "max_speed",
    double_t,
    0,
    "Specifies the max possible speed for the car.",
    20,
    0.0001,
    20
)

gen.add(
    "min_possible_steering_angle",
    double_t,
    0,
    "Specifies the min possible steering value (0, 0.99) that the car should be able to steer at every speed.",
    0.25,
    0.0001,
    0.99
)

gen.add(
    "max_steering_angle",
    double_t,
    0,
    "Max steering angle of the car in degrees.",
    20,
    0.0001,
    50
)

gen.add(
    "pid_p",
    double_t,
    0,
    "Specifies the p of the PID-Controller.",
    2.0,
    0.0001,
    10
)

gen.add(
    "pid_i",
    double_t,
    0,
    "Specifies the i of the PID-Controller.",
    1.0,
    0.0001,
    10
)

gen.add(
    "pid_d",
    double_t,
    0,
    "Specifies the d of the PID-Controller.",
    0.02,
    0.0001,
    10
)

gen.add(
    "anti_windup",
    double_t,
    0,
    "Specifies the anti_windup of the PID-Controller.",
    0.2,
    0.0001,
    10
)

exit(gen.generate(PACKAGE, "wallfollowing5", "wallfollowing5"))
